### ISSUE TITLE
Octave.app won't open after hang in rundemos() or other hangs

### ISSUE THREAD

**Issue by SystemsSage** [octave-app/octave-app#10]
Octave froze while running test demo. (Sorry, can't say where).

Now, it won't start (see attached). I tried removing it and resintallling. Same problem. 

Mac OS 10.13.5. JDK Version 8 Update 191 (build 1.8.0_191-b12) - later version then specified, but was already installed.

![octave error](https://user-images.githubusercontent.com/2261149/49333519-afe94500-f596-11e8-852e-fd49c8aa2439.png)

**Comment by apjanke** [comment]
Uh oh, that's not good.

Which test demo were you running? Was it the test suite run with `__run_test_suite__`?

> JDK Version 8 Update 191 (build 1.8.0_191-b12) - later version then specified, but was already installed.

That shouldn't be a problem. It will mean that Java is disabled for your Octave (since it requires the exact specified version), but it should degrade gracefully and stuff should still run.

**Comment by SystemsSage** [comment]
All I can remember is that it was a demo script run through the command window. guidemo??

Sorry, I wasn’t expecting the problem and did pay enough attention.

**Comment by apjanke** [comment]
No worries. I've got a macOS 10.13 machine; I'll fiddle around with it when I have some time and try to reproduce.

> BTW an uninstall script would have been helpful :-)

Ideally, it shouldn't be necessary: the normal way to uninstall Mac applications is to just delete them from /Applications, and that ought to work for Octave.app too. Though maybe we do need one to completely wipe all the settings files from the machine and give you a completely fresh start. Since you're still having trouble starting it, that suggests some leftover state files interfering with it. But I don't know what would cause that.

If you want to completely get rid of Octave.app's support files, do this at the Terminal:

```
rm -rf ~/.octave_* "$HOME/Library/Application Support/Octave.app"
```

That will remove all the settings/state files that I know of. Let's see if that makes a difference?

**Comment by SystemsSage** [comment]
After cleaning out the various Octave remnants, I was able to install Octave again and it would open. 

The file I ran was "rundemos".
Problems started here:

------------ output chunk #1 ------------

__- test questdlg with five inputs
Press <enter> to continue:
/Applications/Octave-4.4.0.app/Contents/Resources/usr/Cellar/octave-octave-app_4.4.0/4.4.0/share/octave/4.4.0/m/gui/uibuttongroup.m example 1:
 f = figure;
 gp = uibuttongroup (f, "Position", [ 0 0.5 1 1], ...
                     "selectionchangedfcn", ...
                     @(x, y) display (['Selection Changed: ' get(y.NewValue, 'String')]));
 b1 = uicontrol (gp, "style", "radiobutton", ...
                 "string", "Choice 1", ...
                 "Position", [ 10 150 100 50 ]);
 b2 = uicontrol (gp, "style", "radiobutton", ...
                 "string", "Choice 2", ...
                 "Position", [ 10 50 100 30 ]);
 b3 = uicontrol (f, "style", "radiobutton", ...
                 "string", "Not in the group", ...
                 "Position", [ 10 50 100 50 ]);
 disp (['Current selected: ' get(get(gp, 'selectedobject'), 'String')]);
 pause (0.5);
 disp (['Select None']);
 set (gp, 'selectedobject', []);
 pause (0.1);
 disp (['Current selected: ' get(get(gp, 'selectedobject'), 'String')]);
 pause (0.5);
 disp (['Select b1']);
 set (gp, 'selectedobject', b1);
 disp (['Current selected: ' get(get(gp, 'selectedobject'), 'String')]);

warning: implicit conversion from numeric to char
Current selected:
Select None
warning: implicit conversion from numeric to char
Current selected:
Select b1
Current selected: Choice 1
Press <enter> to continue:
/Applications/Octave-4.4.0.app/Contents/Resources/usr/Cellar/octave-octave-app_4.4.0/4.4.0/share/octave/4.4.0/m/gui/uigetdir.m example 1:
 uigetdir (pwd, 'Select Directory');_

objc[4937]: Class FIFinderSyncExtensionHost is implemented in both /System/Library/PrivateFrameworks/FinderKit.framework/Versions/A/FinderKit (0x7fffb42d3c90) and /System/Library/PrivateFrameworks/FileProvider.framework/OverrideBundles/FinderSyncCollaborationFileProviderOverride.bundle/Contents/MacOS/FinderSyncCollaborationFileProviderOverride (0x112188cd8). One of the two will be used. Whichone is undefined.
Press <enter> to continue:
/Applications/Octave-4.4.0.app/Contents/Resources/usr/Cellar/octave-octave-app_4.4.0/4.4.0/share/octave/4.4.0/m/gui/uigetfile.m example 1:
 uigetfile ({'*.gif;*.png;*.jpg', 'Supported Picture Formats'});_

---------- end of output chunk #1 -----------

A file browser opens up. Last time I simple hit return and I suspect that created a problem. This time I created a .png file from a screen grab (project account, no image files). And the demo continued.

Then "rundemos" did this: 

------------ output chunk #2 ------------

Press <enter> to continue:
/Applications/Octave-4.4.0.app/Contents/Resources/usr/Cellar/octave-octave-app_4.4.0/4.4.0/share/octave/4.4.0/m/gui/uimenu.m example 1:
 clf;
 surfl (peaks);
 colormap (copper (64));
 shading ('interp');
 f = uimenu ('label', '&File', 'accelerator', 'f');
 e = uimenu ('label', '&Edit', 'accelerator', 'e');
 uimenu (f, 'label', 'Close', 'accelerator', 'q', 'callback', 'close (gcf)');
 uimenu (e, 'label', 'Toggle &Grid', 'accelerator', 'g', 'callback', 'grid (gca)');

Press <enter> to continue:
/Applications/Octave-4.4.0.app/Contents/Resources/usr/Cellar/octave-octave-app_4.4.0/4.4.0/share/octave/4.4.0/m/gui/uiputfile.m example 1:
 uiputfile ({'*.gif;*.png;*.jpg', 'Supported Picture Formats'});

---------- end of output chunk #2 -----------

No file browser opened. Figure window was open from previous example (2D Surface). When I clicked on Figure window, it cleared. Cannot close Figure window. Pressing the "enter" key simply caused CR in the Command panel. cntl-c would not interrupt (also tried option-c and command-c). File->Quit is greyed out. Cannot close Octave window. 

Only option is to Force Quit the Application through Mac OS. This time, Octave would reopen. Ran "rundemos" again, same problem, same place. 

Hope this helps.

**Comment by apjanke** [comment]
I can reproduce the hang in `rundemos`, but I can't reproduce the not-starting-afterwards problem: if I force-quit Octave.app, launching it again works fine.

The problem appears to be with the `uiputfile()` call that the last demo is doing. If I start up Octave.app and just do this:

```
uiputfile ({'.gif;.png;*.jpg', 'Supported Picture Formats'});
```

That will hang Octave in the same manner, too.

Could you try futzing around with it some more and see if you can reproduce the not-opening-any-more issue? That's more serious than a hang in uiputfile.

I'm going to target this issue for fixing in 4.4.1+, since I suspect this will need to go upstream to Octave, and I don't have time to build up a 4.4.0 non-Octave.app Octave to do a differential test on.

**Comment by apjanke** [comment]
The `uiputfile` hang happens in Octave 4.4.1 for me, too.

**Comment by apjanke** [comment]
Looks like this might be a known upstream bug in Octave: https://savannah.gnu.org/bugs/index.php?52840. I've commented there.

**Comment by SystemsSage** [comment]
Good work. At least we know where the problem likely is.

Gary

Plus ça change, plus c'est la même chose.

**Comment by apjanke** [comment]
Since I can reproduce this in a vanilla Octave 4.4.1 build outside of Octave.app, I think this is just an upstream issue with Octave itself, and not part of the Octave.app packaging. As such, I'm going to close this out as an upstream issue. Please follow the [upstream bug report](https://savannah.gnu.org/bugs/index.php?52840) for updates, and chime in there if you want to add pressure for fixing it.

If this turns out to be specific to Octave.app (e.g. you can get `uiputfile()` to work under a vanilla Octave build with Qt enabled), please re-open.

**Comment by brandonfosso** [comment]
Hi, I am seeing similar behavior from Octave.app under slightly different conditions. I installed 4.4.1 a few hours ago and it worked fine for about 15 min. Then a script I was running for a Coursera class crashed it and it is refusing to come back up. When I try to launch it from spotlight search, the terminal icon appears in the dock, is briefly replaced with the octave icon, then reverts and this message shows up:
![image](https://user-images.githubusercontent.com/7997179/53771510-74695a00-3ea8-11e9-96c2-6d0fc520c7d3.png)
 I am working on this machine:
![image](https://user-images.githubusercontent.com/7997179/53771541-9662dc80-3ea8-11e9-8b77-01bdfc3c14a1.png)

I have tried uninstalling and reinstalling the .app a few times. I even installed 4.4.0 using brew but it has the same behaviour. I tried cleaning the config files with the command you listed above but got the same behavior.

I am not sure where to go next, thanks for your help.

**Comment by brandonfosso** [comment]
Also, before it crashed for good, I did set one preference: I checked the box to have it start in the working directory it was in when it closed.

**Comment by apjanke** [comment]
Could you see if there's a crash log in `~/Library/Logs/DiagnosticReports` and if so, attach it here?

> I tried cleaning the config files with the command you listed above but got the same behavior.

That command actually only delete's Octave.app's own supplementary config files. There are also Octave's own config & state files at `~/.config/octave`. Try blowing them all away with:

```
rm -rf ~/.config/octave ~/.octave_* "$HOME/Library/Application Support/Octave.app"
```

and then restarting Octave.app.

**Comment by brandonfosso** [comment]
Hi, thanks for getting back to me so quickly. There were many crash reports from octave (guess I tried opening it many times) and the latest is attached.

[octave-gui_2019-03-04-180601_Behemoth.crash.txt](https://github.com/octave-app/octave-app/files/2932709/octave-gui_2019-03-04-180601_Behemoth.crash.txt)

Also, I removed the files you mentioned and octave launches just fine now but I can reproduce the behavior simply by checking the preference box to start in the directory it left off in.

Under Preferences > General:
![image](https://user-images.githubusercontent.com/7997179/53835876-1b570000-3f54-11e9-9888-0e153aa9fdd2.png)

**Comment by brandonfosso** [comment]
Anyways, once Octave is open, it works fine but it freezes when I close it. I have to use the macOS Force Quit to put it down. That can't be the expected behavior! ;)

Also, here is the latest crash report related to the checkbox mentioned above:
[octave-gui_2019-03-05-143429_Behemoth.crash.txt](https://github.com/octave-app/octave-app/files/2932759/octave-gui_2019-03-05-143429_Behemoth.crash.txt)

Thanks again for you help, I really appreciate it.

**Comment by brandonfosso** [comment]
Hey, one more thing, I was playing with closing Octave a bit more and I noticed that it exits cleanly when I close it from the command line in the app with 'exit' or 'quit'. But if I try to quit with 'cmd + q' or octave-gui > quit octave-gui it either does nothing other than grey out the menu quit option (below) or sometimes it puts 'exit' into the prompt with an apparent newline (below) but does not quit. In that situation, pressing enter will run the 'exit' command. 

![image](https://user-images.githubusercontent.com/7997179/53837040-35dea880-3f57-11e9-963e-0ab92dd601b3.png)

![image](https://user-images.githubusercontent.com/7997179/53837138-835b1580-3f57-11e9-8adc-c5bb28478697.png)

So to summarize, the macOS ways of quitting don't result in the expected behavior but Octave continues to run and can be closed using 'exit' or 'quit' from its command line.

Thanks again for your help. The program is use-able so it is time for me to get back on that homework. Also, sorry for dumping this all in one thread, lmk if you want me to create a new issue or something.

**Comment by apjanke** [comment]
Oh, that's interesting. ⌘-Q or "octave-gui" > "Quit octave-gui" has never done anything for me. It's always been greyed out and nonresponsive. I just assumed it was something Mac Octave didn't support, and it seemed too minor to complain about at this point.

Thank you for the crash reports and investigation! I'm going to play around with that "Restore working directory of previous session" preference and see if I can get this to reproduce on my box.

**Comment by apjanke** [comment]
> Also, sorry for dumping this all in one thread, lmk if you want me to create a new issue or something.

Let's keep this all in one thread here; this seems relevant to the original bug report. Re-opening.

**Comment by apjanke** [comment]
I've played around with this for a while, and upgraded my desktop to Mojave, and I'm afraid I just can't reproduce it.

Going to leave this issue open since users are experiencing it; hopefully someone else can help to reproduce.

**Comment by apjanke** [comment]
Hey folks. Could you help diagnose this by attaching spindump reports to this issue?

While Octave.app is hung, run Activity Monitor and locate the `octave-gui` process. (It'll probably be labeled as "Not Responding".) Click the gear button in the upper left of Activity Monitor and choose "Spindump". Once the report is produced, in the report window, click the "Save..." button in the upper right, and save it as something like "octave-5.1.0-hang-on-launch.spindump". Then compress that in a .zip file and attach it to a comment here.

**Comment by ozandernek** [comment]
Hi all, 

I have a slightly different problem. I am using oh-my-zsh and octave was working without a problem until I upgraded my system to Mojave. Now when I try to start octave it gives me the error: 

“Octave-4.4.1” cannot be opened because the developer cannot be verified. macOS cannot verify that this app is free from malware.

And there are bunch of crash log at ~/Library/Logs/DiagnosticReport.

**Comment by apjanke** [comment]
@ozandernek: Sounds unrelated.

Check your System Preferences > Security & Privacy > General and make sure that "Allow apps downloaded from:" is set to "App Store and identified developers" and not just "App Store".

Then go in to /Applications, right-click on Octave-4.4.1, and choose "Open". Then it should give you a dialog box asking you if you want to allow it to run. Select Yes/OK, and it should run, and once you've done that once it should run normally like any other app.

**Comment by ozandernek** [comment]
@apjanke: Thanks ! I guess my computer glitched for some reason. Switching that option off and on again solved the problem.

**Comment by apjanke** [comment]
Closing bc I was never able to reproduce this, and no recent updates.

If this is still a problem, please check out the newer Octave.app releases like [8.4.0](https://github.com/octave-app/octave-app/releases/tag/v8.4.0) and [9.1.0 beta4](https://github.com/octave-app/octave-app/releases/tag/v9.1.0_beta4), and let me know if it's still happening there.
